src/graph.c

```diff
@@ -249,6 +251,153 @@ void store_clique(Bitset *clique, Bitset **cliques, uint32_t *num_of_cliques) {
   }
 }
 
+Graph* find_complement_undirected_graph(Graph* graph) {
+  Graph* complement_graph = (Graph*)calloc(1, sizeof(Graph));
+  initialize_graph(complement_graph, graph -> vertices, 0);
+
+  for (int v = 1; v <= graph -> vertices; v++)
+  {
+    for (int u = v + 1; u <= graph -> vertices; u++)
+    {
+      if (!(graph -> adjacency_matrix[v][u] && graph -> adjacency_matrix[u][v]))
+      {
+        add_edge(complement_graph, u, v, 1);
+        add_edge(complement_graph, v, u, 1);
+        complement_graph -> edges += 2;
+      }
+    }  
+  }
+
+  return complement_graph;
+}
+
+Bitset** clique_aprox(Graph* graph, uint32_t* num_of_cliques) {  
+  int** exclude = (int**)calloc(graph -> vertices + 1, sizeof(int*));
+  for (int i = 1; i <= graph -> vertices; i++)
+  {
+    exclude[i] = (int*)calloc(graph -> vertices + 1, sizeof(int));
+  }
+  ASSERT(exclude != NULL, "Could not allocate space for exclude.");
+
+  int** exclude_next = (int**)calloc(graph -> vertices + 1, sizeof(int*));
+  for (int i = 1; i <= graph -> vertices; i++)
+  {
+    exclude_next[i] = (int*)calloc(graph -> vertices + 1, sizeof(int));
+  }
+  ASSERT(exclude_next != NULL, "Could not allocate space for exclude_next.");
+
+  int** cost = (int**)calloc(graph -> vertices + 1, sizeof(int *));
+  for (int i = 1; i <= graph->vertices; i++)
+  {
+    cost[i] = (int*)calloc(graph -> vertices + 2, sizeof(int));
+  }
+  ASSERT(cost != NULL, "Could not allocate space for cost.");
+
+  int** previous = (int**)calloc(graph -> vertices + 1, sizeof(int *));
+  for (int i = 1; i <= graph->vertices; i++)
+  {
+    previous[i] = (int*)calloc(graph -> vertices + 2, sizeof(int));
+  }
+  ASSERT(previous != NULL, "Could not allocate space for previous.");
```

Comment:
Jeśli sprawdzasz czy wskaźnik na początek tablicy jest poprawny, to można również sprawdzić czy wskaźniki na poszczególne wiersze są poprawne.